metersink/output_odoo.py

```diff
@@ -129,157 +236,323 @@ def get_odoo_user_id(odoo):
+        # projection_dict = get_projection_dict(model="sale.order.line")
+
+        # line_records = odoo_get(odoo, 'sale.order.line',
+        #                         mode='read',
+        #                         filter_list=[line_ids],
+        #                         # projection_dict=projection_dict
+        #                         )
+
+        line_record = odoo_get(
+            odoo,
+            "sale.order.line",
+            mode="search_read",
+            filter_list=[
+                [
+                    ["order_id", "=", sale_orders["id"]],
+                    ["product_id", "=", 36],
+                ]
+            ],
+            # projection_dict=projection_dict
+        )
+        if line_record:
+            # todo update
+            # we are reading the last state of the line and calculate the time between this line and the update message.
+
+            if LOG.isEnabledFor(logging.DEBUG):
+                LOG.debug("%s", pformat(line_record))
+
+            time_calc = calculate_cloud_time(line_record["name"])
+
+            quantity = 3000
+            odoo_update(
+                odoo,
+                "sale.order.line",
+                line_record["id"],
+                {"product_uom_qty": quantity},
+            )
+        else:
+            # If there is no line already for the ressource, create it.
+
+            time_calc = calculate_cloud_time(
+                data_dict["traits"]["created_at"]
+            )
+
+            line_dict = {
+                "product_id": 36,
+                "name": display_name,
+                "product_uom_qty": time_calc,
+                "order_id": sale_orders["id"],
+            }
+            line_id = odoo_create(odoo, "sale.order.line", [line_dict])
+            LOG.debug("%s", line_id)
+
+
+def create_sale_order(odoo, customer, tag_list):
+    """creates a new SO for customer with tags"""
+    record_list = [
+                {
+                    "partner_id": customer['id'],
+                    "category_ids": tag_list
+                }
+            ]
+    new_id = odoo_create(odoo, "sale.order", record_list)
+    return new_id
+
+
+def create_sale_order_line(odoo, order_id, product_id, display_name, product_uom_qty):
+    """creates a new so line and returns its id"""
+    record_list = [
+        {
+            "order_id": order_id,
+            "product_id": product_id,
+            "display_name": display_name,
+            "product_uom_qty": product_uom_qty,
+        }
+    ]
+    new_id = odoo_create(odoo, "Sale.order.line", record_list)
+    return new_id
+
+
+def get_sale_order_id(odoo, tag_list):
     """
-    looks for already existing entries.
+    Returns the id of a Sale_order to further work on if the Customer is known.
+    Else it gives nothing back.
     """
-    # print(line)
-    if not filter_dict:
-        filter_dict = {}
+    # project_tag = f"project={project_id}"
+
+    [customer] = odoo_get_customer_from_project(odoo, tag_list)
```

Comment:
If this function always returns a one-element list, maybe it should only return the one element instead?

Reply:
I changed the function so it is more flexible now and makes it more clear why it returns a list.